File: internal/hl7/HL7Field.go
package hl7Messages

import (
	"bytes"
	"strings"
)

type HL7Field struct {
	Delimiters      *MessageDelimiters
	Index           int
	Components      []*HL7Component
	Field           string
	SegmentName     string
	RepititionIndex int
}

func NewHL7Field(index int, field, segmentName string, delimiters *MessageDelimiters, repititionIndex int) *HL7Field {
	return &HL7Field{
		Delimiters:      delimiters,
		Index:           index,
		Components:      []*HL7Component{},
		Field:           field,
		SegmentName:     segmentName,
		RepititionIndex: repititionIndex,
	}
}

func (f *HL7Field) ParseComponents(seperator string) {
	if seperator == "" {
		seperator = f.Delimiters.ComponentSeperator
	}
	index := 1
	for _, component := range strings.Split(f.Field, string(seperator)) {
		if component != f.Field {
			repititionIndex := 0
			for _, repitition := range strings.Split(component, string(f.Delimiters.RepetitionSeperator)) {
				if repitition == component {
					hl7Field := NewHL7Component(index, repitition, f.SegmentName, f.Delimiters, -1)
					hl7Field.ParseComponents()
					f.Components = append(f.Components, hl7Field)
				} else {
					hl7Field := NewHL7Component(index, repitition, f.SegmentName, f.Delimiters, repititionIndex)
					hl7Field.ParseComponents()
					f.Components = append(f.Components, hl7Field)
					repititionIndex++
				}
				index++
			}
		}
	}
}

func (f *HL7Field) GetComponent(index int) *HL7Component {
	for _, c := range f.Components {
		if c.Index == index {
			return c
		}
	}
	return nil
}

func (f *HL7Field) String() string {
	if len(f.Components) > 0 {
		var output bytes.Buffer
		for _, subField := range f.Components {
			if subField.Index > 1 && subField.RepititionIndex <= 0 {
				output.WriteString(string(f.Delimiters.ComponentSeperator))
			} else if subField.RepititionIndex > 0 {
				output.WriteString(string(f.Delimiters.RepetitionSeperator))
			}
			output.WriteString(subField.String())
		}
		return output.String()
	} else {
		return f.Field
	}
}

File: ClientApp/src/app/message/message-models/HL7Field.ts
import { HL7Component, IHL7Field } from './HL7Component'
import { HL7Segment } from './HL7Segment'
import { MessageDelimeters } from './MessageDelimeters'

export class HL7Field {
   public Components: IHL7Field[] = []

   public constructor(
      public Index: number,
      public Field: string,
      public SegmentName: string,
      protected delimiters: MessageDelimeters,
      public RepititionIndex = -1
   ) {}

   ParseComponents(seperator: string = this.delimiters.ComponentSeperator) {
      let index = 1
      HL7Segment.SplitField(this.Field, seperator, this.delimiters.EscapeCharacter).forEach(
         (component) => {
            if (component != this.Field) {
               let repititionIndex = 0
               HL7Segment.SplitField(
                  component,
                  this.delimiters.RepetitionSeperator,
                  this.delimiters.EscapeCharacter
               ).forEach((repitition) => {
                  if (repitition == component) {
                     let hl7Field = new HL7Component(
                        index,
                        repitition,
                        this.SegmentName,
                        this.delimiters
                     )
                     hl7Field.ParseComponents()
                     this.Components.push(hl7Field)
                  } else {
                     let hl7Field = new HL7Component(
                        index,
                        repitition,
                        this.SegmentName,
                        this.delimiters,
                        repititionIndex++
                     )
                     hl7Field.ParseComponents()
                     this.Components.push(hl7Field)
                  }
                  index++
               })
            }
         }
      )
   }

   public ToString() {
      if (this.Components.length) {
         let output: string[] = []
         this.Components.sort(HL7Field.SortFunction).forEach((subField) => {
            if (subField.Index > 1 && subField.RepititionIndex <= 0) {
               output.push(this.delimiters.ComponentSeperator)
            } else if (subField.RepititionIndex > 0) {
               output.push(this.delimiters.RepetitionSeperator)
            }
            output.push(subField.ToString())
         })
         return output.join('')
      } else {
         return this.Field
      }
   }

   protected static SortFunction(a: IHL7Field, b: IHL7Field) {
      if (a.Index > b.Index) {
         return 1
      } else if (a.Index < b.Index) {
         return -1
      } else {
         if (a.RepititionIndex > b.RepititionIndex) {
            return 1
         } else if (a.RepititionIndex < b.RepititionIndex) {
            return -1
         }
         return 0
      }
   }
}
